File: SERIAL/Transpose/transpose.f90
!
! Copyright (c) 2015, Intel Corporation
!
! Redistribution and use in source and binary forms, with or without
! modification, are permitted provided that the following conditions
! are met:
!
! * Redistributions of source code must retain the above copyright
!      notice, this list of conditions and the following disclaimer.
! * Redistributions in binary form must reproduce the above
!      copyright notice, this list of conditions and the following
!      disclaimer in the documentation and/or other materials provided
!      with the distribution.
! * Neither the name of Intel Corporation nor the names of its
!      contributors may be used to endorse or promote products
!      derived from this software without specific prior written
!      permission.
!
! THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
! "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
! LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
! FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
! COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
! INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
! BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
! LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
! LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
! POSSIBILITY OF SUCH DAMAGE.

!*******************************************************************
!
! NAME:    transpose
!
! PURPOSE: This program measures the time for the transpose of a
!          column-major stored matrix into a row-major stored matrix.
!
! USAGE:   Program input is the matrix order and the number of times to
!          repeat the operation:
!
!          transpose <matrix_size> <# iterations> [tile size]
!
!          An optional parameter specifies the tile size used to divide the
!          individual matrix blocks for improved cache and TLB performance.
!
!          The output consists of diagnostics to make sure the
!          transpose worked and timing statistics.
!
!
! FUNCTIONS CALLED!:
!
!          Other than standard C functions, the following
!          functions are used in this program:
!
!          wtime()          portable wall-timer interface.
!
! HISTORY: Written by  Rob Van der Wijngaart, February 2009.
!          Converted to Fortran by Jeff Hammond, January 2015
! *******************************************************************

program main
  use iso_fortran_env
  implicit none
  ! for argument parsing
  integer :: err
  integer :: argnum, arglen
  character(len=32) :: argtmp
  ! problem definition
  integer(kind=INT32) ::  iterations                ! number of times to do the transpose
  integer(kind=INT32) ::  order                     ! order of a the matrix
  real(kind=REAL64), allocatable ::  A(:,:)         ! buffer to hold original matrix
  real(kind=REAL64), allocatable ::  B(:,:)         ! buffer to hold transposed matrix
  integer(kind=INT64) ::  bytes                     ! combined size of matrices
  ! runtime variables
  integer(kind=INT32) ::  i, j, k
  integer(kind=INT32) ::  it, jt, tile_size
  real(kind=REAL64) ::  abserr, addit, temp         ! squared error
  real(kind=REAL64) ::  t0, t1, trans_time, avgtime ! timing parameters
  real(kind=REAL64), parameter ::  epsilon=1.D-8    ! error tolerance

  ! ********************************************************************
  ! read and test input parameters
  ! ********************************************************************

  print*,'Parallel Research Kernels version ', -1 !PRKVERSION
  print*,'Serial Matrix transpose: B = A^T'

  if (command_argument_count().lt.2) then
    print*,'argument count = ', command_argument_count()
    print*,'Usage: ./transpose <# iterations> <matrix order> [<tile_size>]'
    stop 1
  endif

  iterations = 1
  call get_command_argument(1,argtmp,arglen,err)
  if (err.eq.0) read(argtmp,'(i)') iterations

  order = 1
  call get_command_argument(2,argtmp,arglen,err)
  if (err.eq.0) read(argtmp,'(i)') order

  tile_size = order
  if (command_argument_count().gt.2) then
      call get_command_argument(3,argtmp,arglen,err)
      if (err.eq.0) read(argtmp,'(i)') tile_size
  endif

  if (iterations .lt. 1) then
    print*,'ERROR: iterations must be >= 1 : ', iterations
    stop 1
  endif

  if (order .lt. 1) then
    print*,'ERROR: order must be >= 1 : ', order
    stop 1
  endif

  if ((tile_size .lt. 1).or.(tile_size.gt.order)) then
    print*,'WARNING: tile_size must be >= 1 and <= order : ', tile_size
    tile_size = order ! no tiling
  endif


  ! ********************************************************************
  ! ** Allocate space for the input and transpose matrix
  ! ********************************************************************

  allocate( A(order,order), stat=err)
  if (err .ne. 0) then
    print*,'allocation of A returned ',err
    stop 1
  endif

  allocate( B(order,order), stat=err )
  if (err .ne. 0) then
    print*,'allocation of A returned ',err
    stop 1
  endif

  ! avoid overflow 64<-32
  bytes = 2 * order
  bytes = bytes * order

  print*,'Matrix order = ', order
  print*,'Number of iterations = ', iterations
  print*,'Tile size = ', tile_size

  ! Fill the original matrix, set transpose to known garbage value. */

  !  Fill the original column matrix
  do j=1,order
    do i=1,order
      A(i,j) = (i-1)+(j-1)*order
    enddo
  enddo

  !   Set the transpose matrix to a known garbage value.
  do j=1,order
    do i=1,order
      B(i,j) = 0.0
    enddo
  enddo

  do k=0,iterations

    !  start timer after a warmup iteration
    if (k.eq.1) call cpu_time(t0)

    !  Transpose the  matrix; only use tiling if the tile size is smaller than the matrix
    if (tile_size.lt.order) then
      do j=1,order,tile_size
        do i=1,order,tile_size
          do jt=j,min(order,j+tile_size-1)
            do it=i,min(order,i+tile_size-1)
              B(jt,it) = B(jt,it) + A(it,jt)
              A(it,jt) = A(it,jt) + 1.0
            enddo
          enddo
        enddo
      enddo
    else
      do j=1,order
        do i=1,order
          B(j,i) = B(j,i) + A(i,j)
          A(i,j) = A(i,j) + 1.0
        enddo
      enddo
    endif

  enddo ! iterations

  ! ********************************************************************
  ! ** Analyze and output results.
  ! ********************************************************************

  call cpu_time(t1)
  trans_time = t1 - t0

  abserr = 0.0;
  addit = (0.5*iterations)
  addit = addit * (iterations+1)
  do j=1,order
    do i=1,order
      ! this was overflowing for iterations>15
      temp   = ((j-1.)+(i-1.)*order)
      temp   = temp * (iterations+1)
      abserr = abserr + abs(B(i,j) - (temp+addit))
    enddo
  enddo

  deallocate( B )
  deallocate( A )

  if (abserr .lt. epsilon) then
    print*,'Solution validates'
    avgtime = trans_time/iterations
    print*,'Rate (MB/s): ',1.e-6*bytes/avgtime, &
           ' Avg time (s): ', avgtime
    stop
  else
    print*,'ERROR: Aggregate squared error ',abserr, &
           'exceeds threshold ',epsilon
    stop 1
  endif

end program main

File: SERIAL/Transpose/transpose.c
/*
Copyright (c) 2013, Intel Corporation

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

* Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
* Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.
* Neither the name of Intel Corporation nor the names of its
      contributors may be used to endorse or promote products
      derived from this software without specific prior written
      permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
*/

/*******************************************************************

NAME:    transpose

PURPOSE: This program measures the time for the transpose of a
         column-major stored matrix into a row-major stored matrix.

USAGE:   Program input is the matrix order and the number of times to
         repeat the operation:

         transpose <matrix_size> <# iterations> [tile size]

         An optional parameter specifies the tile size used to divide the
         individual matrix blocks for improved cache and TLB performance.

         The output consists of diagnostics to make sure the
         transpose worked and timing statistics.


FUNCTIONS CALLED:

         Other than standard C functions, the following
         functions are used in this program:

         wtime()          portable wall-timer interface.

HISTORY: Written by  Rob Van der Wijngaart, February 2009.
         Modernized by Jeff Hammond, February 2016.
*******************************************************************/

#include <par-res-kern_general.h>

int main(int argc, char ** argv)
{
  /*********************************************************************
  ** read and test input parameters
  *********************************************************************/

  printf("Parallel Research Kernels version %s\n", PRKVERSION);
  printf("Serial Matrix transpose: B = A^T\n");

  if (argc != 4 && argc != 3) {
    printf("Usage: %s <# iterations> <matrix order> [tile size]\n", *argv);
    exit(EXIT_FAILURE);
  }

  int iterations  = atoi(*++argv); /* number of times to do the transpose */
  if (iterations < 1) {
    printf("ERROR: iterations must be >= 1 : %d \n", iterations);
    exit(EXIT_FAILURE);
  }

  int order = atol(*++argv); /* order of a the matrix */
  if (order < 0) {
    printf("ERROR: Matrix Order must be greater than 0 : %d \n", order);
    exit(EXIT_FAILURE);
  }

  int tile_size = 32; /* default tile size for tiling of local transpose */
  if (argc == 4) {
      tile_size = atoi(*++argv);
  }
  /* a non-positive tile size means no tiling of the local transpose */
  if (tile_size <=0) {
      tile_size = order;
  }

  /*********************************************************************
  ** Allocate space for the input and transpose matrix
  *********************************************************************/

  size_t bytes = (size_t)order * (size_t)order * sizeof(double);

  double (* const restrict A)[order] = (double (*)[order]) prk_malloc(bytes);
  if (A == NULL){
    printf(" Error allocating space for transposed matrix\n");
    exit(EXIT_FAILURE);
  }
  double (* const restrict B)[order] = (double (*)[order]) prk_malloc(bytes);
  if (B == NULL){
    printf(" Error allocating space for input matrix\n");
    exit(EXIT_FAILURE);
  }

  printf("Matrix order          = %d\n", order);
  if (tile_size < order) {
      printf("Tile size             = %d\n", tile_size);
  } else {
      printf("Untiled\n");
  }
  printf("Number of iterations  = %d\n", iterations);

  /*  Fill the original matrix, set transpose to known garbage value. */

  /* Fill the original column matrix. */
  {
      for (int j=0;j<order; j++) {
        for (int i=0;i<order; i++) {
          A[j][i] = (double) ((size_t)order*(size_t)j+(size_t)i);
        }
      }
  }

  /*  Set the transpose matrix to a known garbage value. */
  {
      for (int j=0;j<order; j++) {
        for (int i=0;i<order; i++) {
          B[j][i] = 0.0;
        }
      }
  }

  double trans_time = 0.0;

  for (int iter = 0; iter<=iterations; iter++){

    /* start timer after a warmup iteration */
    if (iter==1) trans_time = wtime();

    /* Transpose the  matrix */
    if (tile_size < order) {
      for (int it=0; it<order; it+=tile_size) {
        for (int jt=0; jt<order; jt+=tile_size) {
          for (int i=it; i<MIN(order,it+tile_size); i++) {
            for (int j=jt; j<MIN(order,jt+tile_size); j++) {
              B[i][j] += A[j][i];
              A[j][i] += 1.0;
            }
          }
        }
      }
    } else {
      for (int i=0;i<order; i++) {
        for (int j=0;j<order;j++) {
          B[i][j] += A[j][i];
          A[j][i] += 1.0;
        }
      }
    }
  }  /* end of iter loop  */

  trans_time = wtime() - trans_time;

  /*********************************************************************
  ** Analyze and output results.
  *********************************************************************/

  double addit = ((double)(iterations+1) * (double) (iterations))/2.0;
  double abserr = 0.0;
  for (int j=0;j<order;j++) {
    for (int i=0;i<order; i++) {
      abserr += ABS(B[j][i] - ((double)(order*(size_t)i+(size_t)j)*(iterations+1L)+addit));
    }
  }

  prk_free(B);
  prk_free(A);

#ifdef VERBOSE
  printf("Sum of absolute differences: %f\n",abserr);
#endif

  const double epsilon = 1.e-8;
  if (abserr < epsilon) {
    printf("Solution validates\n");
    double avgtime = trans_time/iterations;
    printf("Rate (MB/s): %lf Avg time (s): %lf\n", 1.0E-06 * (2L*bytes)/avgtime, avgtime);
#ifdef VERBOSE
    printf("Squared errors: %f \n", abserr);
#endif
    exit(EXIT_SUCCESS);
  }
  else {
    printf("ERROR: Aggregate squared error %lf exceeds threshold %e\n", abserr, epsilon);
    exit(EXIT_FAILURE);
  }

  return 0;
}


